thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/api/detection/AnomalyDetectionResource.java

```diff
@@ -0,0 +1,756 @@
+   * @param principal user who sent this request. It's used to separate different config names
+   * @return a message containing the detected anomalies and the detection config used
+   */
+  @POST
+  @Path("/")
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.APPLICATION_JSON)
+  @ApiOperation("Request an anomaly detection online task")
+  public Response onlineApi(
+          @QueryParam("start") long start,
+          @QueryParam("end") long end,
+          @ApiParam("jsonPayload") String payload,
+          @Auth ThirdEyePrincipal principal) {
+    DatasetConfigDTO datasetConfigDTO = null;
+    MetricConfigDTO metricConfigDTO = null;
+    DetectionConfigDTO detectionConfigDTO = null;
+    TaskDTO taskDTO = null;
+    List<AnomalySummary> anomalies = null;
+    Response.Status responseStatus;
+    Map<String, String> responseMessage = new HashMap<>();
+    ObjectMapper objectMapper = new ObjectMapper();
+    // Use username to separate different requests. One user can only send one request at a time
+    String nameSuffix = "_" + principal.getName();
+
+    try {
+      if (payload.getBytes().length > MAX_ONLINE_PAYLOAD_SIZE) {
+        responseStatus = Response.Status.BAD_REQUEST;
+        responseMessage.put("message", "Payload too large");
+        return Response.status(responseStatus).entity(responseMessage).build();
+      }
+
+      JsonNode payloadNode = objectMapper.readTree(payload);
+
+      if (!validateOnlineRequestPayload(payloadNode)) {
+        responseStatus = Response.Status.BAD_REQUEST;
+        responseMessage.put("message", "Invalid request payload");
+        return Response.status(responseStatus).entity(responseMessage).build();
+      }
+
+      // Preprocess: remove existing entities generated by the previous interrupted request
+      cleanExistingOnlineTask(nameSuffix);
+
+      // Create & save dataset
+      datasetConfigDTO = generateDatasetConfig(payloadNode, nameSuffix);
+
+      // Create & save metric along with online data
+      metricConfigDTO = generateMetricConfig(payloadNode, nameSuffix);
+
+      // Create & save detection
+      detectionConfigDTO =
+          generateDetectionConfig(payloadNode, nameSuffix, datasetConfigDTO, metricConfigDTO, start,
+              end);
+
+      // Create & save task
+      taskDTO = generateTaskConfig(detectionConfigDTO.getId(), start, end);
+
+      // Polling task status
+      TaskDTO polledTaskDTO = pollingTask(taskDTO.getId());
+
+      // Task failure
+      if (polledTaskDTO.getStatus() != TaskConstants.TaskStatus.COMPLETED) {
+        LOG.warn("Task is not completed after polling: " + polledTaskDTO);
+
+        responseStatus = Response.Status.INTERNAL_SERVER_ERROR;
+
+        switch (polledTaskDTO.getStatus()) {
+        case FAILED:
+          responseStatus = Response.Status.INTERNAL_SERVER_ERROR;
+          responseMessage.put("message", "Failed to execute anomaly detection task.");
```

Comment:
What about attaching the failure stacktrace here rather than just a generic error message, since the error might help users to debug their configurations.

Reply:
Yes. Same as above, I already put that after this switch statement. Thanks!